Add examples to the `Fun` module

hyphenrf opened the pull request:
In a similar spirit to #11476, I tried to add examples of varying complexity to the combinators in `Stdlib.Fun`.
The convention I chose to follow was the one agreed on in the aforementioned PR.

In addition to examples, I also noted what I think may be "gotchas", namely the evaluation of `const`'s argument and using `flip` with n > 2 args functions.

Will add a changes entry if this is deemed a worthwhile change. Thanks in advance to kind reviewers!

Commit:
Add examples to the `Fun` module

hyphenrf commented:
A prototype of these examples was present in the `Fun.on` PR, which Gabriel gave me some tips on, and I applied them here.

Octachron commented on stdlib/fun.mli:
```diff
@@ -61,3 +63,67 @@ exception Finally_raised of exn
+
+    One common mode of usage for {!combinators} is to create and pass anonymous
+    functions as arguments to higher-order functions, without needing to specify
+    [(fun ... -> ...)] explicitly.
```
Personally, I never use combinators to avoid writing an explicit lambda. Factorizing common patterns is a much better motivation from my point of view (which tends to exclude toy combinators like  `flip` and `curry` or `uncurry`).

hyphenrf replied:
> Factorizing common patterns

could you please expand on this a bit more? I can add it to the passage or just be more neutral and snip this part

Octachron replied:
Maybe the focus on anonymous function is not necessary either:

> One common mode of usage for {!combinators} is to create short-lived arguments for higher-order functions

Or maybe we should emphasis the lightweight idea rather constrasting with abstraction:

>  {{!combinators}Combinators} provide a lightweight way to create short-lived arguments for higher-order functions

Octachron commented on stdlib/fun.mli:
```diff
@@ -61,3 +63,67 @@ exception Finally_raised of exn
+    [(fun _ -> ...)] directly is that the expression denoted [(...)] will be
+    evaluated once, and the function [const] returns will only have access to
+    the result of evaluating [(...)]. So really, [const (...)] is not equivalent
+    to [(fun _ -> ...)], but rather to [let result = ... in (fun _ -> result)].
```
The fact that `const` evalutate its argument once is a remark. I would advise to describe first what  `const` is before describing finer use points. 

Moreover, the paragraph feels too long and verbose. I would argue mentioning that `const` evaluates its argument once while using the (nice) `Random.bool`  example to illustrate the point should be enough.

hyphenrf replied:
I don't want to repeat what's written in the `const` doc comment which I feel does describe what `const` is.. What do you have in mind? Provide an example implementation `let const x _ = x` maybe?

I'm also worried the bool example on its own wouldn't be too apparent if one doesn't have a strong mental model of how things evaluate already.. I want to at least keep the equivalence fact so that people can experiment with it.

Octachron replied:
You are right that re-explaining `const` would be verbose. Part of the issue is maybe trying to explain `const` out of context. If we apply the idea to have a thematic `building a Seq.t`, there is probably no need to do so.

My fear is that the equivalence explanation increases the probability that readers will not read the paragraph nor the example. We should keep in mind that the example should target the audience that skipped the function definition.

I would propose at least to start with a short sentence: `beware "const" evaluates its argument once`, then the `Random` example and its explanation. I am not sure if we need to write down the equivalence.

Octachron commented on stdlib/fun.mli:
```diff
@@ -61,3 +63,67 @@ exception Finally_raised of exn
+    ]}
+    Note that given a function [(f : a -> b -> c -> d)]:
+    - [(flip f x : a -> c -> d)] and [(x : b)], whereas
+    - [(flip (f x) : c -> b -> d)] and [(x : a)]
```
Honestly, I am (as expected) unconvinced by those examples, and I fear that trying to find  good examples for `flip` is a wild good chase. I would propose to split this case to another PR if you want to argue for it in more details.

hyphenrf replied:
I looked up uses in OCaml code out there, and In most cases `flip` is used
- to match arg order with another function.. e.g. the `List.rev` example from here.. which I could replace with a `Set.add` example at the risk of more code just causing obscurity,
- to change datalast to datafirst.. e.g. to filter a list by which elements being members of a set: `filter (flip S.mem s)`.. somewhat related to the first case,
- or to change the arg order of a binary operator where the order has semantic meaning.. e.g. the `flip compare`  and `subtract` examples.

Octachron commented on stdlib/fun.mli:
```diff
@@ -61,3 +63,67 @@ exception Finally_raised of exn
+    - [(flip f x : a -> c -> d)] and [(x : b)], whereas
+    - [(flip (f x) : c -> b -> d)] and [(x : a)]
+
+    {!val:negate}
```
It might work better to have a thematic `predicate` section which could contain example of `negate`, `const true` and `const false`.

hyphenrf replied:
There's value in having one overarching theme in the examples, I agree, but I feel it'll take a bit of power out of the other examples. WDYT? I'll try to come up with nicer examples to the other three following the predicate theme...

Octachron replied:
Maybe we could start the `negate` section a short description of predicates as functions of type `'some -> bool` often used in `List.find_all` and similar functions. Then `Fun.negate` could be presented as quick (aka without eta-expansion) way to reverse a predicate.

hyphenrf replied:
Forgot to reply here, I added the passage:
```
{3:hnegate {{!val:negate}negate}}

Mainly used for reversing a predicate in a function which expects one, like
{!val:List.find_all} and similar functions
```
I opted not to explain the word "predicate", relying on the precedence in `negate`'s own doc comment, and leaving space for the explanation to be put there instead.

Octachron commented on stdlib/fun.mli:
```diff
@@ -61,3 +63,67 @@ exception Finally_raised of exn
+    [(fun ... -> ...)] explicitly.
+
+    The examples below will be demonstrating this mainly with the {!module:List}
+    module.
```
I am wondering if `Seq` would not be a better illustration. In particular, we could have a section on building sequences to illustrate `const` and `id` and `Seq` has more sequence constructor like `forever` or `ints` to use as a basis of more examples.

hyphenrf replied:
The problem with using `Seq`, answering also your comment on `const` above, is in three aspects I imagine:

- it's less straightforward due to its delay semantics, and its potentially infinite nature.. Instead of focusing on one function, there's a potential to add more distractions (`List.of_seq`, `Seq.take`, `Seq.equal`...)
- it's less common in code and pedagogical material. list manipulation is one of the first things beginners work through
- it doesn't play as nicely with the toplevel evaluation theme we have going on, because `Seq.t` prints as `<fun>`

Commit:
Reword some paragraphs in `Fun` examples

Commit:
Reword more paragraphs in `Fun` examples

Commit:
Add more examples to `Fun.id` and `const`

hyphenrf commented:
Changes done since the last review:
- The prologue and the `const` paragraphs are shortened, more neutral wording in the prologue and equivalence in const removed.
- `flip` with n > 2 ary functions is slightly expanded on
- More examples are added to `id` and `const`, particularly the IMO elegant `List.filter_map Fun.id` replaces `List.find_all Fun.id` which felt artificial.For `const` I opted for a somewhat thought provoking example that follows from `const id`, here used in tandem with `fold_left`.

hyphenrf commented:
```
Interesting equivalences:
   - [const id] = [flip const] = [fun _ y -> y]
   - [flip id] = [(|>)] = [fun x f -> f x] (this is because [(|>)] is
     [flip (@@)], and [(@@)] is a specialization of [id : 'a -> 'a] where ['a]
     is unified with [('b -> 'c)])
```
Thinking about adding this at the bottom. Is it a good idea?

NoahTheDuke commented:
i'd love to see this merged, i think it's good

gasche commented on stdlib/fun.mli:
```diff
@@ -61,3 +63,81 @@ exception Finally_raised of exn
+    # List.init 3 (Fun.const 0);;
+    - : int list = [0; 0; 0]
+
+    # let last xs = List.fold_left (Fun.const Option.some) None xs;;
```
I find this example obscure -- applying Fun.const to a one-element to a two-element function is weird. I would write it `fun _acc x -> Some x` or something. The `List.init` example above makes more sense to me.

hyphenrf replied:
Quick ack & response until I can re-examine the code with fresh eyes: the idea was to progress into more "fun" examples to give the user a feel of the capabilities of those functions, with the most straightforward example being the first. I could just keep one straightforward example for each function, two at most, and avoid this delights style entirely.

Octachron replied:
I would propose the new `String.spellcheck` function as a natural use case for `Fun.const`.

hyphenrf replied:
Addressed by removing examples that are too obscure in terms of combinator usage, and indicating new "advanced" examples which try to be straightforward in terms of combinator usage, but more elaborate in utility. Also added caution paragraphs on highlighted observations/equivalences

hyphenrf commented:
@NoahTheDuke Thanks for reminding me to resume work on this. I was deliberating on whether or not to split into one PR for each function, but I feel like in doing so we might achieve a good example set in one of them that compromises the rest when the document is considered in full. I'm trying to nail the "overarching theme" as Octachron suggested. I want to avoid local maximums so I'm hoping to finish the discussion here. I'm working on something unrelated currently that's eating up most of my day. When I get a chance in a day or two from now, I should also update the PR to include examples for `compose` :)

Octachron commented on stdlib/fun.mli:
```diff
@@ -61,3 +63,81 @@ exception Finally_raised of exn
+
+    # let to_flat = Float.Array.map_from_array Fun.id
+    val to_flat : float array -> Float.Array.t
+    ]}
```
I think it might make sense to add an one sentence description to each examples.

hyphenrf replied:
Added descriptions

Commit:
Remove problematic examples and add descriptions

Commit:
Expand the Fun.id dispatching description

And make the example conform to the description

Commit:
Add "advanced"/"real-world" examples

These are instead of the problematic examples

Commit:
Make function refs headers

Commit:
Add `Fun.compose` examples

And warn about too many combinators

Commit:
Deindent and improve formatting

hyphenrf commented:
It turned out to be a month not a day or two, apologies!
I've addressed the points raised in reviews and added examples for `compose`.

The PR should preferably be reviewed as one document at this point,
a) because I admit my commits turned out not so atomic, especially when I noticed I needed to deindent the document or the code blocks turn out indented when seen through `make html_doc` output, but also
b) because the examples are meant to be readable in one sitting, in the style of the previous referenced PR, so it's valuable to consider the document as a whole, with coherence/continuity as a relation between the examples, not just consider each examples on its own. Some examples reference and build on previous examples even.

I should add here that, at this stage, I'm open to reviewers requesting whole examples and paragraphs removed outright if they deem them redundant or weakening etc. I've added a bit of variety and I'm satisfied with each example. I'm not too attached to any single example as long as the maintainers judge the set they choose to keep meets the desired level of understanding and showcase of utility.

Notably, I've added cautions regarding the value restriction and regarding readability, to guide the reader against going too crazy with these functions.
At the end of that, I wasn't sure how to crossref the manual, and didn't want to reference it by URL until I know if that's an OK thing to do in stdlib docs.

I'd really appreciate a second pair of eyes on the generated document, once the maintainers have approved/selected the content. Suggestions for better prose, grammatical mistakes, and the overall aesthetic with element choices are very welcome. I have written a good amount of documentation, but I'm no ocamldoc guru. There was a best-effort attempt to follow best practices nonetheless.

hyphenrf commented:
~~CI is complaining about a Changes entry, but do pure doc contributions get an entry?~~ (apparently they do)
Will add a changes entry last.

hyphenrf commented:
(Actually, let me think through `compose` examples a bit more. I think something with the `Format` API could be more worthwhile to consider as an example. I'll work more on this PR. I'll also take a second look at the prose in general. I want to shrink the addition a bit and be more focused I think.)

Interestingly for `compose`, the PR which introduced it (#12770) seems like it was merged without the author making much of a case for the combinator. So I couldn't take a shortcut and use the author's case they would've made in a back-and-forth discussion as an example here.

I personally don't use compose much if at all because it's not an operator, so multiple compositions don't improve readability, and writing out `compose` in full more often than not turns out to be less concise and more obscure than expanding it. (This is as opposed to an `on` combinator which this PR was originally a split of, I find myself needing that one from time to time, as it at least removes duplication). You could even try that with the examples I provided, expanding compose makes them more concise and/or more readable.

I did consider an example which shows a pipeline of `List.map`s, talks about the inefficiency of constructing intermediates like that, then proposes a single `map` of `compose`d functions as an alternative, but that falls short for `n ≠ 2` maps, again because `compose` is not an operator, and seems more like a thing where converting to `Seq` instead may be encouraged. I'm very hesitant to define an operator for `compose` in the docs, because that may implicitly bless it, and everyone has their own operator for it. <sub>Maybe in a future where `let ( ∘ ) ...` is a valid syntax, that will stop being the case, who knows</sub>

Perhaps a fun way to use compose is for difference list appending, but I didn't want to introduce that whole concept. It's not a data-structures tutorial. <sub>Perhaps if one day in the far future difference lists are included in the stdlib, compose would make a cameo in _its_ docs. :grin: </sub>

Speaking of the compose examples, these are common use-cases of compose, but perhaps the combinator is useful in something more specific to us? It was a bit hard for me to dig through sherlocode comprehensively for this one, because it's often an operator, and doesn't have "standard" notation (the prior PRs acknowledge that fact). What I did dig out showed pretty much what I included here: selectors like `snd`, and data transformation. If someone would like to point me to something in the wild where compose has a nice utility, I'd be happy to include it instead of one of or both of the examples. What do you find yourself using `compose` for?

gasche commented:
> do pure doc contributions get an entry? 

Yes, of course!

gasche commented on stdlib/fun.mli:
```diff
@@ -67,3 +69,199 @@ exception Finally_raised of exn
+as it is depending on the {!val:Sys.win32} value
+{[
+if Sys.win32 then String.map (function '/' -> '\\' | c -> c) else Fun.id
+]}
```
I'm not fond of this example: it is long enough that it makes sense to `let-bind` it, and then I would more naturally write it in eta-expanded form `let normalize str = if Sys.win32 then ... else str`. 

I've seen use-cases as you describe in the following situations, where eta-expanding is less natural:
1. The value returned by a conditional or pattern-matching contains a wrapper function, but also other things.
2. The wrapper function is passed as an argument to a function instead of being returned.

An example of (1) from the OCaml compiler codebase:

https://github.com/gasche/ocaml/blob/3393613e7e0d60ad0a9c9ac63bed9f06232b8ddb/asmcomp/cmmgen.ml#L833-L838

An example of (2):

https://github.com/gasche/ocaml/blob/3393613e7e0d60ad0a9c9ac63bed9f06232b8ddb/asmcomp/dataflow.ml#L27

hyphenrf replied:
I opted for the second situation, and included a simple logger function example instead of the path (de)normalization

hyphenrf commented:
Few questions after stepping back and thinking a bit more about the document:
1. Should the examples be convincing, as in, make you want to use the function? Or is it enough that they showcase how the function may be used, even if they turn out obviously artificial? This could just be shrunk to five simple examples, one for each function, and that be it.
2. Should the prose be in tutorial style, like it currently is? Or should it be more matter-of-fact and make less suggestions?
3. Is thematic consistency a worthy goal to pursue?
4. Should we be limited to REPL-style blocks? Or is pseudo-syntax equations/equivalences style better suited?
5. Is linking urls (e.g. https://ocaml.org/manual/5.3/polymorphism.html#ss:valuerestriction) in stdlib docs a good idea?
6. Is it a good idea to introduce data structures like set, map, or difference lists, which would require the user to do more work to reproduce the example?

For me, 1. I'd like strong examples yes, 2. Seems reasonable to guide against future friction, 3. indifferent, 4. It's certainly more flexible, but it's the style used in the doc comments of each function itself. Examples should be the concrete aid to the abstract equation, 5. unsure but depends on 2 anyway, 6. against

gasche commented on stdlib/fun.mli:
```diff
@@ -67,3 +69,199 @@ exception Finally_raised of exn
+let rec chain = function
+  | [] -> Fun.id
+  | f :: fs -> fun x -> f (chain fs x)
+]}
```
Another advanced example is as a final continuation for functions in continuation-passing-style:

```ocaml
type 'a btree =
  | Leaf of 'a
  | Node of 'a btree * 'a * 'a btree

let rec map_cps f tree k =
  match tree with
  | Leaf v -> k (Leaf (f v))
  | Node (left, v, right) ->
    map_cps f left @@ fun left' ->
    let v' = f v in
    map_cps f right @@ fun right' ->
    Node (left', v', right')

let map f tree = map_cps f tree Fun.id
```

hyphenrf replied:
That was originally in place of `chain`, but I wasn't sure if introducing the concept of CPS in an example was a case worth defending

gasche commented on stdlib/fun.mli:
```diff
@@ -67,3 +69,199 @@ exception Finally_raised of exn
+{3:hflip {{!val:flip}flip}}
+
+Use [flip] to reverse the comparator passed to {!val:List.sort}, resulting in a
+reversed sorting
```
I find "reversed sorting" a bit strange, I would maybe write "which sorts in decreasing order:" or "which sorts in the opposite order:".

gasche commented on stdlib/fun.mli:
```diff
@@ -67,3 +69,199 @@ exception Finally_raised of exn
+- : int list = [3; 3; 2]
+]}
+
+From the {{!hconst}[spellcheck] example}, [flip] could've been used instead of
```
```suggestion
From the {{!hconst}[spellcheck] example}, [flip] could have been used instead of
```

gasche commented on stdlib/fun.mli:
```diff
@@ -67,3 +69,199 @@ exception Finally_raised of exn
+of two lists (with order and duplicates from the second)
+{[
+# List.find_all (Fun.flip List.mem [2; 3; 5]) [0; 3; 3; 2; 4; 6; 8]
+- : int list = [3; 3; 2]
```
I don't really like this example and the one below. They use the form `flip f x` to partially-apply the second argument of `f`, which I think is too clever, I would rather write an explicit function in this case.

hyphenrf replied:
That was what I was going for, `flip` with the second argument partially applied. I don't mind removing it.
The rationale was that it was common enough (you see it in Set, Map, List, Stack, Queue, ... wherever you want to make datafirst become datalast or the opposite)

Update: it is removed

gasche commented on stdlib/fun.mli:
```diff
@@ -67,3 +69,199 @@ exception Finally_raised of exn
+
+{!val:List.find_all} string elements with length of exactly 3
+{[
+# List.find_all (Fun.compose ((=)3) String.length) ["one"; "two"; "three"]
```
```suggestion
# List.find_all (Fun.compose ((=) 3) String.length) ["one"; "two"; "three"]
```

gasche commented on stdlib/fun.mli:
```diff
@@ -67,3 +69,199 @@ exception Finally_raised of exn
+val negate : ('a -> bool) -> 'a -> bool
+]}
+
+From the {{!hid}[chain] example}, [compose] could've been used in the recursive
```
```suggestion
From the {{!hid}[chain] example}, [compose] could have been used in the recursive
```

Commit:
Reword some descriptions

Commit:
Remove problematic `flip` examples

Commit:
Replace a `compose` example with more involved one

gasche reviewed (approved):
My overall impression of this category of examples is rather positive. This is a non-trivial, thoughtful contribution, I think it may be of interest to some readers; and it does an appreciated effort of warning about the sins of point-free programming. It may be a good discipline in the future when considering other combinators to ask for similar convincing examples and grow this documentation accordingly.

I propose to wait to see if other people want to give a second read and give another round of feedback. (Ping me in two weeks as I will forget to check.) Then I would propose to go ahead and merge.

Commit:
Typos in the Uchar example

Commit:
Some rewording

Commit:
Replace the windows path conversion example

Commit:
Add a Scanf example to id

Commit:
Add example to const, remove example from compose

Commit:
Indent code blocks, reorder, reformat, reword

Commit:
Remove scanf and floatarray examples from id

hyphenrf commented:
So I couldn't leave this for a while and instead continued work on it haha

I think I'm happy with the example set right now. I chose to shrink some sections to improve the focus in them. I'm ready for reviews and open to example additions (especially a second REPL example in compose... maybe...)

The examples I removed, for reference:
```ocaml
(* id *)
Scanf.scanf "%d" Fun.id

Float.Array.map_from_array Fun.id

if Sys.win32 then String.map (function '/' -> '\\' | c -> c) else Fun.id


(* const *)
let last xs = List.fold_left (Fun.const Option.some) None xs


(* flip *)
let subtract = Fun.flip (-) in
List.map (subtract 2) [4; 6; 8]

List.find_all (Fun.flip List.mem [2; 3; 5]) [0; 3; 3; 2; 4; 6; 8]

(* a flip reference of a present const example *)
let spellcheck known_words =
   String.spellcheck ~max_dist:(Fun.const 2) (Fun.flip List.iter known_words)


(* compose *)
List.find_all (Fun.compose ((=) 3) String.length) ["one"; "two"; "three"]

let buf = Buffer.create 16 in
List.iter (Fun.compose (Buffer.add_utf_8_uchar buf) Uchar.of_int) [0x49; 0x2764; 0xfe0f; 0x1f42b];
Buffer.contents buf
```

gasche commented:
@hyphenrf I would like to move to merge this, but for this we need two things:

- there should be a Changes entry; documentation contributions are also important
- our `check-typo` script complains that fun.mli line 226 has 81 columns, you need to break lines to fit in 80 columns.

Commit:
Prevent overlong lines in code examples

Commit:
Add a Changes entry

hyphenrf commented:
Done (Also updated the branch because I had some local issues with the shallow repo I cloned and originally based this branch on, sorry for the noise).

I'm struggling with rendering the docs in a modern format (the `make` target I invoke ends up creating old-style docs) and I just need confirmation the style is consistent and follows convention by anyone who knows their way around the doc generation pipeline better...

For reference, this is exactly what I did to generate the docs:
```
./configure
make core
make html_doc
xdg-open ./api_docgen/ocamldoc/build/html/libref/Fun.html
```
![image](https://github.com/user-attachments/assets/a487ce1f-e0d4-44cf-afb6-816964d0df9a)

Commit:
Merge branch 'trunk' into hazem/fun-examples

gasche commented:
Merged! We haven't had a final look by documentation-building experts, but I'm impatient to get the nice contribution in so I merged anyway. If there are minor rendering glitches we can always fix them later.

hyphenrf commented:
Well, I'm thankful for the patience with the reviews :grin:
Thanks all
